server/src/internalClusterTest/java/org/elasticsearch/reservedstate/service/FileSettingsServiceIT.java

```diff
@@ -381,6 +384,40 @@ public void testErrorSaved() throws Exception {
         assertClusterStateNotSaved(savedClusterState.v1(), savedClusterState.v2());
     }
 
+    public void testLastSettingsInBatchApplied() throws Exception {
+        internalCluster().setBootstrapMasterNodeIndex(0);
+        logger.info("--> start data node / non master node");
+        String dataNode = internalCluster().startNode(Settings.builder().put(dataOnlyNode()).put("discovery.initial_state_timeout", "1s"));
+        FileSettingsService dataFileSettingsService = internalCluster().getInstance(FileSettingsService.class, dataNode);
+
+        assertFalse(dataFileSettingsService.watching());
+
+        logger.info("--> start master node");
+        final String masterNode = internalCluster().startMasterOnlyNode();
+        assertMasterNode(internalCluster().nonMasterClient(), masterNode);
+        var savedClusterState = setupClusterStateListener(masterNode);
+
+        FileSettingsService masterFileSettingsService = internalCluster().getInstance(FileSettingsService.class, masterNode);
+
+        assertTrue(masterFileSettingsService.watching());
+        assertFalse(dataFileSettingsService.watching());
+
+        final var masterNodeClusterService = internalCluster().getCurrentMasterNodeInstance(ClusterService.class);
+        final var barrier = new CyclicBarrier(2);
+        masterNodeClusterService.createTaskQueue("block", Priority.NORMAL, batchExecutionContext -> {
+            safeAwait(barrier);
+            safeAwait(barrier);
+            batchExecutionContext.taskContexts().forEach(c -> c.success(() -> {}));
+            return batchExecutionContext.initialState();
+        }).submitTask("block", ESTestCase::fail, null);
+
+        safeAwait(barrier);
+        writeJSONFile(masterNode, testJSON, versionCounter, logger);      // Valid but skipped
+        writeJSONFile(masterNode, testJSON43mb, versionCounter, logger);  // The last valid setting
```

Comment:
Could we sometimes try 3, I have a hunch that might cover some logic that the 2 case doesn't.

Could we give them different `version` values? And not always apply them in ascending order?

Could we wait for the master to pick up one file before writing the next?

Reply:
Good ideas!

Waiting for the master to pick up one file before writing the next would exercise existing functionality, rather than my changes, but it's obviously an important case to test.

Reply:
What is the expected behaviour when the versions are not ascending?

Should `ReservedStateUpdateTaskExecutor` be attempting them in descending version order rather than just iterating backward through the batch list?

Reply:
The guiding principle is that the expected outcome should be the same whether the tasks are executed in a single batch or one-at-a-time. So yes AIUI I think we should attempt them in descending version order.

Reply:
Ah of course, that makes perfect sense. I'll make that change.

Reply:
(Again, it seems like we're going to a lot of trouble to simulate running them in order, instead of actually running them in order.)

Reply:
You have to call `success` (or `onFailure`) to record the outcome of the task, but that doesn't mean you have to actually do anything else with the task. According to the guiding principle, skipping a task because we processed a newer one counts as success right?

Reply:
Alright makes sense. So perhaps:

1. Scan the tasks to identify the one that would have run last, had the tasks all been run individually instead of in a batch, taking into consideration task ordering, version numbers, and the `ReservedStateVersionCheck` mode.
2. Try that task. If it succeeds, call `success` on that one and all prior ones, and possibly on all following ones?
4. If it fails, call `onFailure`, remove it from the candidates list, and loop back to #1.

Reply:
Yeah that sounds right although maybe it'd be simpler just to sort the list of tasks in the right order ?

Reply:
I was afraid I couldn't determine the right order before I know whether the first fails. But now that I think about it, if the first task succeeds, the rest are irrelevant, so I could sort them on the assumption that each one fails.